File: src/governance.c
#include "../include/bcind_kernel.h"

int load_governance_state(GovernanceState* state, const char* filepath) {
    if (state == NULL || filepath == NULL) {
        return -1;
    }

    FILE* fp = fopen(filepath, "r");
    if (!fp) {
        // Defaults if file doesn't exist yet
        state->state_id = 1;
        state->execution_permitted = true;
        state->immutable_lock = true;
        strncpy(state->jurisdiction, "IEC62304", sizeof(state->jurisdiction)-1);
        state->audit_sequence = 1000;
        return save_governance_state(state, filepath);
    }

    char buffer[512];
    size_t bytes_read = fread(buffer, 1, sizeof(buffer)-1, fp);
    fclose(fp);
    buffer[bytes_read] = '\0';

    // Basic JSON state extraction
    state->state_id = 1;
    state->execution_permitted = (strstr(buffer, "\"execution_permitted\": true") != NULL || strstr(buffer, "\"execution_permitted\":true") != NULL);
    state->immutable_lock = (strstr(buffer, "\"immutable_lock\": true") != NULL || strstr(buffer, "\"immutable_lock\":true") != NULL);
    
    char* jur_ptr = strstr(buffer, "\"jurisdiction\":");
    if (jur_ptr) {
        char jur_val[16] = {0};
        if (sscanf(jur_ptr, "\"jurisdiction\": \"%15[^\"]\"", jur_val) == 1) {
            strncpy(state->jurisdiction, jur_val, sizeof(state->jurisdiction)-1);
        }
    } else {
        strncpy(state->jurisdiction, "IEC62304", sizeof(state->jurisdiction)-1);
    }

    char* seq_ptr = strstr(buffer, "\"audit_sequence\":");
    if (seq_ptr) {
        unsigned long seq = 0;
        if (sscanf(seq_ptr, "\"audit_sequence\": %lu", &seq) == 1 || sscanf(seq_ptr, "\"audit_sequence\":%lu", &seq) == 1) {
            state->audit_sequence = (uint64_t)seq;
        } else {
            state->audit_sequence = 1000;
        }
    } else {
        state->audit_sequence = 1000;
    }

    return 0;
}

int save_governance_state(const GovernanceState* state, const char* filepath) {
    if (state == NULL || filepath == NULL) {
        return -1;
    }

    FILE* fp = fopen(filepath, "w");
    if (!fp) {
        return -1;
    }

    fprintf(fp, "{\n");
    fprintf(fp, "  \"state_id\": %u,\n", state->state_id);
    fprintf(fp, "  \"execution_permitted\": %s,\n", state->execution_permitted ? "true" : "false");
    fprintf(fp, "  \"immutable_lock\": %s,\n", state->immutable_lock ? "true" : "false");
    fprintf(fp, "  \"jurisdiction\": \"%s\",\n", state->jurisdiction);
    fprintf(fp, "  \"audit_sequence\": %lu\n", (unsigned long)state->audit_sequence);
    fprintf(fp, "}\n");

    fclose(fp);
    return 0;
}

File: src/governance.rs
use serde::{Deserialize, Serialize};
use std::fs::File;
use std::io::{Read, Write};
use std::path::Path;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct GovernanceState {
    pub state_id: u32,
    pub execution_permitted: bool,
    pub immutable_lock: bool,
    pub jurisdiction: String,
    pub audit_sequence: u64,
}

impl Default for GovernanceState {
    fn default() -> Self {
        Self {
            state_id: 1,
            execution_permitted: true,
            immutable_lock: true,
            jurisdiction: "IEC62304".to_string(),
            audit_sequence: 1000,
        }
    }
}

pub fn load_governance_state<P: AsRef<Path>>(
    filepath: P,
) -> Result<GovernanceState, std::io::Error> {
    let path = filepath.as_ref();
    if !path.exists() {
        let default_state = GovernanceState::default();
        save_governance_state(&default_state, path)?;
        return Ok(default_state);
    }

    let mut file = File::open(path)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;

    let state: GovernanceState = serde_json::from_str(&contents)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
    Ok(state)
}

pub fn save_governance_state<P: AsRef<Path>>(
    state: &GovernanceState,
    filepath: P,
) -> Result<(), std::io::Error> {
    let path = filepath.as_ref();
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent)?;
    }
    let mut file = File::create(path)?;
    let contents = serde_json::to_string_pretty(state)
        .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
    file.write_all(contents.as_bytes())?;
    Ok(())
}
